File: AoC-2022/Day6/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <deque>
#include <set>

using namespace std;

void find_message(string& msg, int mark_len)
{
  deque<char> deq;
  set<char> cset;

  for(auto i = 0; i<mark_len; ++i)
    deq.push_back(msg[i]);
    
  for(auto i = mark_len; i < msg.size(); ++i)
  {
    for(auto i : deq)
      cset.insert(i);

    if(cset.size() == mark_len)
    {
      cout<<"Solution: "<<i<<endl;
      break;
    }
    cset.clear();
    deq.pop_front();
    deq.push_back(msg[i]);
  }
}

int main()
{
  ifstream file("input.txt");
  stringstream stream;
  stream << file.rdbuf();
  string message = stream.str();
  
  find_message(message, 4);
  find_message(message, 14);

  return 0;
}

File: AoC-2022/Day6/solution.py
from collections import deque

def find_message(f, msg_len):
  deq = deque()
  for i in range(0,msg_len):
    deq.append(f[i])

  for i in range(msg_len, len(f)):
    if (len(set(deq)) == msg_len):
      print(i)
      break
    deq.popleft()
    deq.append(f[i])

file = open('input.txt').read()
find_message(file, 4)
find_message(file, 14)